### PR TITLE
plugins: implement Sanitize UGen

### PR DESCRIPTION
**Pull request by snappizz**
A few open questions with this one.

Do we also need `SimpleNumber:sanitize`? How should it be implemented?

Should we add support for post output? I chose to leave it out since it seemed like too much duplication of functionality with CheckBadValues — use CheckBadValues if you want to debug, and use Sanitize if you want to stop worrying about bad values altogether. Another option is to duplicate the `post` option and have it off by default.

### CODE REVIEW

**Comment by telephon** [comment]
>Should we add support for post output?

no, this is excellent.

A primitive for sclang could be added if necessary, but this seems fine for now.

**Comment by brianlheim** [comment]
This is excellent. Just wanted to say I'm against `SimpleNumber-sanitize`. inf and NaN are certainly "bad" values in the context of signal processing but outside of that I don't like the implication that these values are dirty. A method that replaces special int/float values with an arbitrarily chosen constant seems like something that the user should implement themself. fwiw I'd prefer `ReplaceBadValues` purely for similarity with `CheckBadValues`.

OT but it would also be nice to have `SimpleNumber-isInf`. There's not a simple method to test for infinity afaik but maybe I missed it.

**Comment by snappizz** [comment]
The problem is, the behavior of the sanitize method is pretty unintuitive in the absence of a client-side implementation:

    { DC.kr(1/0).sanitize }.plot(0.01) // works
    { (1/0).sanitize }.plot(0.01) // doesn't work

Note that `{ DC.ar(1/0).sanitize }.plot(0.01)` doesn't work for some reason, so I have to debug that...

**Comment by telephon** [comment]
the only reason to have the same method for SimpleNumber is that the three classes `UGen`, `SimpleNumber` and `Array` can be used interchangeably as UGen input.

so you could write

`x.sanitize` without having to know what `x` is.

File: server/plugins/TestUGens.cpp [hunk 1]
```diff
@@ -44,16 +44,18 @@ struct CheckBadValues : public Unit
 	int		prevclass;
 };
 
+struct Sanitize : public Unit { };
+
 // declare unit generator functions
-extern "C"
```
Comment (vivid-synth) [inline server/plugins/TestUGens.cpp]: Why isn't the ```extern "C"``` needed?

Comment (snappizz) [inline server/plugins/TestUGens.cpp]: I don't know the difference, but I trust Tim: https://github.com/supercollider/example-plugins/commit/166c18cf3de457731618cb4f8c4bb613e6fab398#commitcomment-17393581

Comment (vivid-synth) [inline server/plugins/TestUGens.cpp]: Hmm I'd love to know why (although I do believe you that it works!) - @timblechmann is ```extern "C"``` specified elsewhere? I thought we need that for un-mangled names?

**Comment by vivid-synth** [comment]
This is going to be really nice to have! Should we add an argument for what value we replace the bad values with, with a default of 0? I can imagine wanting something else.

**Comment by snappizz** [comment]
I suppose so. In that case, Brian's suggested name ReplaceBadValues might be better, although I'm not a fan of the method name `.replaceBadValues`.

**Comment by vivid-synth** [comment]
I'd still be fine with the name "Sanitize". In general people will want the default ```.sanitize``` behavior, but there's an arg if someone wants it.

**Comment by brianlheim** [comment]
I see the point about parallelism among UGen, SimpleNumber, and Array, but
seeing as there's no SimpleNumber-checkBadValues I think the consistency
argument could go in either direction. This seems to me like an inherently
signal-oriented method.

Definitely agree it should take an argument for the replacement value--I
can imagine cases where I would want the output to be 1 instead of 0.

On Sun, Nov 27, 2016 at 5:07 PM vivid-synth <notifications@github.com>
wrote:

I'd still be fine with the name "Sanitize". In general people will want the
default .sanitize behavior, but there's an arg if someone wants it.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/supercollider/supercollider/pull/2514#issuecomment-263151193>,
or mute the thread

**Comment by snappizz** [comment]
Implementation of the replace argument required introducing multiple calculation functions, which introduces more potential for bugs. Here are a few test cases to work from regarding different input rates:

```
{ Sanitize.ar(Select.ar(LFPulse.ar(1000), [DC.ar(1/0), DC.ar(-1)]), Sweep.ar * 100) }.plotAudio(0.01)
{ Sanitize.ar(Select.ar(LFPulse.ar(1000), [DC.ar(1/0), DC.ar(-1)]), Sweep.kr * 100) }.plotAudio(0.01)
{ Sanitize.kr(Select.kr(LFPulse.kr(200), [1/0, -1]), Sweep.ar * 100) }.plotAudio(0.01)
{ Sanitize.kr(Select.kr(LFPulse.kr(200), [1/0, -1]), Sweep.kr * 100) }.plotAudio(0.01)
```

Here's a test to ensure that infinity is properly zeroed out:

```
{ Sanitize.ar(DC.ar(1/0)) }.loadToFloatArray(10 / s.sampleRate, action: _.postln);
```

I had trouble creating test cases for NaN and denormals, so some help there would be appreciated.

**Comment by brianlheim** [comment]
Awesome, thanks. Here are more test case snippets:

`DC.ar(0.01).pow(19)` gives me a denormal, and
`DC.ar(0)/0`, 0/0 give NaN.

**Comment by adcxyz** [comment]
Hi all, 
the SafetyNet quark also has an implementation of this idea (after long discussion), 
ReplaceBadValues: 
http://new-supercollider-mailing-lists-forums-use-these.2681727.n2.nabble.com/New-quark-SafetyNet-td7626835.html#a7627234

It allows replacement of signals instead of hardcoding 0 (which is useful for control signals)
It would make sense to consider integration/coordination here.

best adc

File: HelpSource/Classes/Sanitize.schelp [hunk 2]
```diff
@@ -1,15 +1,19 @@
 class:: Sanitize
-summary:: Zero out infinity, NaN, and denormals
+summary:: Remove infinity, NaN, and denormals
 categories:: UGens>Info
 
 description::
-Replaces infinities, NaNs, and subnormal numbers (denormals) with zeros. The method link::Classes/UGen#-sanitize:: provides a shorthand for this.
+Replaces infinities, NaNs, and subnormal numbers (denormals) with a value of your choice, usually zero. The method link::Classes/UGen#-sanitize:: provides a shorthand for this.
```
Comment (vivid-synth) [inline HelpSource/Classes/Sanitize.schelp]: s/usually/by default/ ?

Comment (brianlheim) [inline HelpSource/Classes/Sanitize.schelp]: I'd write "with a given value, zero by default."

**Comment by vivid-synth** [comment]
Is this ready to merge now?

**Comment by snappizz** [comment]
Do we consider `SimpleNumber:sanitize` a requirement for merge? If not, all this needs is a rebase.

**Comment by snappizz** [comment]
I would also love to get a unit test for this going, but I'm waiting on #2414.

**Comment by vivid-synth** [comment]
I personally don't consider the SimpleNumber work a requirement for this. Agreed on unit tests.

**Comment by telephon** [comment]
>This seems to me like an inherently signal-oriented method. 

The thing is that when you do not write your signal graph by hand, but generate it by some combinatorial method, it may well happen that what is passed to `sanitize` ends up to be a float. 

Even in some cases, UGens are automatically factored out  for optimization. This is why a SimpleNumber method makes sense.

But for me, this isn't a requirement for the current PR.

**Comment by telephon** [comment]
is there anything in the `SafetyNet` Quark that might be worth considering for this UGen?

**Comment by adcxyz** [comment]
Two things: 
it can substitute a signal when replacing, which I find useful for krs such as filter frequencies, 
and its ugen shortcut is nicely short: .zap.
Using the Safety class on the main outputs was heavily requested by users, 
so maybe it is a candidate for later inclusion - but no need to tie it to this PR.

**Comment by telephon** [comment]
>it can substitute a signal when replacing

That means a changing signal (defined by any UGen) ?

Sorry, I guess I should just go and try …

**Comment by adcxyz** [comment]
yes, when its input becomes illegal, it switches to 0 or an optional substitute signal. 
this will keep e.g. filters from exploding with a filterfreq or bandwidth of 0, 
and maybe designed to signal to the player that something is wrong without going silent.

**Comment by telephon** [comment]
ok, that seems like a valuable feature. Maybe that could be added on top once this PR is merged?

**Comment by telephon** [comment]
Also  `zap` seems a little better than `sanitize`, but this is a matter of hygiene culture I guess …

**Comment by adcxyz** [comment]
zapping puts the electricity back into electronic music ;-)

**Comment by jamshark70** [comment]
> This seems to me like an inherently
signal-oriented method.

Suppose somebody wanted to write:

```
Pbind(
    \a, patternThatCouldNaN.sanitize,
    ...
)
```

**Comment by brianlheim** [comment]
> Suppose somebody wanted to write:
> ```
> Pbind(
>     \a, patternThatCouldNaN.sanitize,
>     ...
> )

I'd probably tell them to go back and figure out why it NaNs in the first place! Hahaha. But yeah, I do see now that it makes sense to extend this to SimpleNumber for the reasons @jamshark70 and @telephon have mentioned.

**Comment by adcxyz** [comment]
yes of course you wants to find it, but when you have a show in 30 minutes,
or when you are trying unusual combinations and want to protect your ears and speakers, 
.zap is better than nothing.
2c adc

**Comment by jamshark70** [comment]
In other words, treat `sanitize` like a unary operator and implement it consistently:

- Numbers
- AbstractFunction (composeUnaryOp)
- UGen
- Collections
- Symbol (yes, that sounds weird, but a possibly valid use case is to sanitize an array containing an outgoing OSC message).
- Did I miss any?

**Comment by brianlheim** [comment]
Can you explain the Symbol one a little more?

I was going to suggest Stream but I see it's a subclass of AbstractFunction, so never mind. I think that list is good.

**Comment by adcxyz** [comment]
As Symbol.help says ;-) : 

Math

Symbols respond to all unary and binary math operations by returning themselves. The result of any math operation between a Number or other math object and a Symbol is to return the Symbol. This allows for example operations on lists of notes which contain 'rest's to preserve the rests.

Pseq([1, 3, \rest, 2, 4] + 8);

**Comment by brianlheim** [comment]
Gotcha. In that case, maybe it should be defined on Signal, too?

**Comment by snappizz** [comment]
Signal is a Collection.

**Comment by brianlheim** [comment]
I'm aware of that, but many math ops are defined on the Signal class itself.

**Comment by vivid-synth** [comment]
Maybe we can merge this and create a separate ticket for the sclang class work?

**Comment by snappizz** [comment]
We're waiting for me to rebase, which I'll get around to doing soon.

**Comment by snappizz** [comment]
rebased: #2561

**Pull request closed without merging**